packages/lit-html/src/lib/lit-html.ts

```diff
@@ -12,9 +12,557 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
-export {}
+// Added to an attribute name to mark the attribute as bound so we can find
+// it easily.
+const boundAttributeSuffix = '$lit$';
+
+// This marker is used in many syntactic positions in HTML, so it must be
+// a valid element name and attribute name. We don't support dynamic names (yet)
+// but this at least ensures that the parse tree is closer to the template
+// intention.
+const marker = `lit$${String(Math.random()).slice(9)}$`;
+
+// String used to tell if a comment is a marker comment
+const markerMatch = '?' + marker;
+
+// Text used to insert a comment marker node. We use processing instruction
+// syntax because it's slightly smaller, but parses as a comment node.
+const nodeMarker = `<${markerMatch}>`;
+
+const d = document;
+
+// Creates a dynamic marker. We never have to search for these in the DOM.
+const createMarker = () => d.createComment('');
+
+// https://tc39.github.io/ecma262/#sec-typeof-operator
+type Primitive = null | undefined | boolean | number | string | symbol | bigint;
+const isPrimitive = (value: unknown): value is Primitive =>
+  value === null || !(typeof value === 'object' || typeof value === 'function');
```

Comment:
I had been sticking to strict equality at all times to be conservative and ensure no unneccessary type coercions or ToBoolean operations, and also stuck with `typeof x === 'object'` to ensure we hit the typeof optimization - in case the optimization is fragile.

I checked the bytecode generated in Node for `typeof value === 'object'` vs `typeof value == 'object'` vs `typeof value != 'object'`. Strict and loose equals generate the exact same `TestTypeOf` bytecode, and loose not-equals only adds a logical not. Removing the JS logical not actually removes a ToBoolean as well.

So this is a great suggestion. Thanks for prodding me to look into this more! I've been meaning to.